### PR TITLE
Objects concept

### LINKED ISSUE #107: Objects concept
tasks

- [x] select an exercise
- [x] add concept and exercise to track config.json
- [x] edit concept and exercise config.json
- [x] write tests, exemplar and stub
- [x] edit instructions
- [x] edit design.md
- [x] write about.md
- [x] copy and prune to create introduction.md
- [x] edit hints
- [x] add links

### PR DESCRIPTION
**Pull request by glennj**
Closes #107

### CODE REVIEW

File: concepts/objects/about.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+# About
+
+To use terminology from other languages, a JSON object is a "hash" or "map" or "dictionary".
```
Comment (IsaacG) [inline concepts/objects/about.md:3]:
```suggestion
A JSON object is, to use terminology from other languages, a "hash", "map", or "dictionary".
```

I would start the article with the subject.

File: concepts/objects/about.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+> An object is an unordered set of **name**/**value** pairs.
+> An object begins with `{`left brace and ends with `}`right brace.
+> Each _name_ is followed by `:`colon and the _name/value_ pairs are separated by `,`comma.
+
+Further, the _name_ must be a string.
+Another word for _name_ is _key_.
+
+The _value_ can be of any JSON type.
+Different keys in the object can have values of different type, like this example.
```
Comment (IsaacG) [inline concepts/objects/about.md:15]:
```suggestion
Different _values_ in the same object can be different type, like this example.
```

Using "keys" and "values" both in the sentence makes it a bit confusing about which you're talking about.

File: concepts/objects/about.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+$ echo "Jane" | jq -Rc '{(.): 42}'
+{"Jane":42}
+```
+
+### Short-hand notations
+
+A couple of frequent cases for object construction have short-hand syntax.
+
+1. Re-using key-value pairs from an input object.
```
Comment (IsaacG) [inline concepts/objects/about.md:62]: I see consistent line endings prior to code blocks :smile:

File: concepts/objects/about.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+Given an object that contains array or object values, we can "chain" index expressions together.
+
+```jq
+{
+  "name": "Jane",
+  "age": 42,
+  "pets": ["cat", "fish"],
+  "address": {"street": "123 Main St", "city": "Springfield"}
+} as $example
+|
```
Comment (IsaacG) [inline concepts/objects/about.md:128]: Would this example be simpler without using variables?

File: concepts/objects/about.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+  $example.pets[1],       # => "fish"
+  $example.address.city   # => "Springfield"
+```
+
+## Adding or changing key-value pairs
```
Comment (IsaacG) [inline concepts/objects/about.md:133]: Should "merge" be in the header?

File: concepts/objects/about.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+{Richard: 54} + {Jane: 42}
+# => {
+#      "Richard": 54,
+#      "Jane": 42
+#    }
+```
+
+## Removing a key
+
+Use the `del` function.
```
Comment (IsaacG) [inline concepts/objects/about.md:158]:
```suggestion
Use the `del` function to remove a key.
```

A handful of sections start with sentences that are continuations of the header without being fully formed on their own.

File: concepts/objects/about.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+It returns the updated object.
+
+```jq
+{name: "Jane", age: 42} | del(.age)   # => {name: "Jane"}
+```
+
+## Membership
+
+- Test if the object `has` a key.
```
Comment (IsaacG) [inline concepts/objects/about.md:167]:
```suggestion
- To test if the object has a key, use the `has` function.
```

File: concepts/objects/about.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+- Test if a key is `in` an object
+
+  ```jq
+  {name: "Jane", age: 42} as $example
+  |
+    ("name"  | in($example)),    # => true
+    ("sport" | in($example))     # => false
+  ```
+
+## List all the keys
```
Comment (IsaacG) [inline concepts/objects/about.md:185]:
```suggestion
## List the keys or values
```

File: concepts/objects/about.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+Use the `keys` function to output a list of all the keys.
```
Comment (IsaacG) [inline concepts/objects/about.md:187]: Is it worth adding a `keys` example?

File: concepts/objects/about.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+  ```
+
+- To iterate over an object, we must first convert it to an array of key-value objects.
+  The `to_entries` function does that.
+
+  ```jq
+  {name: "Jane", age: 42} | to_entries'
+  ```
+  outputs
+  ```json
```
Comment (IsaacG) [inline concepts/objects/about.md:217]:
```suggestion
  ```

  outputs

  ```json
```

File: concepts/objects/about.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+  At this point, we can use the array iteration functions, like `map`.
+
+- The `from_entries` function is the inverse: convert an array of key-value objects into an object
+
+  ```jq
+  [
+    {"key":"name", "value":"Jane"},
+    {"key":"age", "value":42}
+  ] | from_entries                  # =>{"name": "Jane", "age": 42}
```
Comment (IsaacG) [inline concepts/objects/about.md:238]:
```suggestion
  ] | from_entries                  # => {"name": "Jane", "age": 42}
```

File: concepts/objects/about.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+  ```
+
+- To apply a filter to _each_ key-value pair in an object, use the `with_entries(filter)` function.
+
+  For example, with an object that maps a name to an age, to swap the keys and values:
```
Comment (IsaacG) [inline concepts/objects/about.md:243]:
```suggestion
  For example, given an object that maps a name to an age, the the key and value can be swapped as follows.
```

File: exercises/concept/high-score-board/.docs/instructions.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
+# Instructions
+
+In this exercise, you are implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
+
+You have 6 functions to implement, mostly related to manipulating an object that holds high scores.
+
+## 1. Create a new high score board
+
+Create a function `create_score_board` that creates an object that serves as a high score board.
```
Comment (IsaacG) [inline exercises/concept/high-score-board/.docs/instructions.md:9]:
```suggestion
Write a function `create_score_board` which creates an object that serves as a high score board.
```
Reduce "stutter"

File: exercises/concept/high-score-board/.docs/instructions.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+The keys of this object will be the names of the players, the values will be their scores.
+For testing purposes, you want to directly include one entry in the object.
+This initial entry should consist of `The Best Ever` as player name and `1000000` as score.
```
Comment (IsaacG) [inline exercises/concept/high-score-board/.docs/instructions.md:12]:
```suggestion
This initial entry should consist of `"The Best Ever"` as player name and `1000000` as score.
```

File: exercises/concept/high-score-board/.docs/instructions.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+| add_player("Dave Thomas"; 0)
+# => {"Dave Thomas": 0, "José Valim": 486373} -- in some order
+```
+
+## 3. Remove players from a score board
+
+If players violate the rules of the arcade hall, they are manually removed from the high score board.
+Implement `remove_player` which takes a board as input and one parameter, the name of the player to remove.
+This function should remove the entry for the given player from the board and output the new board.
+If the player was not on the board in the first place, nothing should happen to the board, it should be returned as is.
```
Comment (IsaacG) [inline exercises/concept/high-score-board/.docs/instructions.md:36]:
```suggestion
If the player was not on the board in the first place, nothing should happen to the board; it should be returned as is.
```

File: exercises/concept/high-score-board/.meta/design.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
+# Design
+
+## Learning objectives
+
+- What are objects - for now describe them as maps/dictionaries that hold key-value pairs, more advanced things will be taught later
```
Comment (IsaacG) [inline exercises/concept/high-score-board/.meta/design.md:5]:
```suggestion
- What are objects. For now describe them as maps/dictionaries that hold key-value pairs, more advanced things will be taught later
```

File: concepts/objects/about.md (hunk 3 above)
Comment (glennj) [reply concepts/objects/about.md:62]: Lesson learned!

File: concepts/objects/about.md (hunk 4 above)
Comment (glennj) [reply concepts/objects/about.md:128]: Do you mean something like this
```sh
$ cat example.json
{
  "name": "Jane",
  "age": 42,
  "pets": ["cat", "fish"],
  "address": {"street": "123 Main St", "city": "Springfield"}
}
$ jq '.pets[1]' example.json        # => "fish"
$ jq '.address.city' example.json   # => "Springfield"
```

File: concepts/objects/about.md (hunk 5 above)
Comment (glennj) [reply concepts/objects/about.md:133]: Hmm, I dunno.

I see that there's not a "changing" example, where I assign to an existing key.

File: concepts/objects/about.md [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+If you cannot refer to the key as an identifier, use _bracket notation_.
+
+```jq
+"name" as $key | {name: "Jane", age: 42} | .$key    # => error
+"name" as $key | {name: "Jane", age: 42} | .[$key]  # => "Jane"
+```
+
+If the key is not in the object, the result is `null`
```
Comment (IsaacG) [inline concepts/objects/about.md:109]:
```suggestion
If the key is not in the object, the result is `null`.
```

File: concepts/objects/about.md (hunk 4 above)
Comment (IsaacG) [reply concepts/objects/about.md:128]:
```
{
  "name": "Jane",
  "age": 42,
  "pets": ["cat", "fish"],
  "address": {"street": "123 Main St", "city": "Springfield"}
}
| (
  .pets[1],
  .addresss.city
)
```

**Review by IsaacG — APPROVED** [review]

**Review by IsaacG — APPROVED** [review]

**Pull request merged**